feat(platform): add documentation for scoped API tokens (#1148)
apify/apify-docs: sources/platform/integrations/programming/api.md

![Integrations page in Apify Console](../images/api-token.png)

> :::caution Warning
> By default, tokens can access all data in your account. If you plan to share the token with a 3rd party, consider [limiting the permission scope](#limited-permissions) of the token, so that it can only access what it really needs.
> :::

Reviewer: ```suggestion
:::caution Token access scope

By default, tokens have full access to your account data. For third-party integratns, consider [limiting the token's permissions](#limited-permissions).

:::
```

## Authentication

You can authenticate the Apify API in two ways. You can either pass the token via the `Authorization` HTTP header or the URL `token` query parameter. We always recommend you use the authentication via the HTTP header as this method is more secure.

> :::caution Important
> Do not share the API token with untrusted parties, or use it directly from client-side code,
> unless you fully understand the consequences!
> :::

Reviewer: ```suggestion
:::caution Secure token usage

Never share your API token with untrusted parties or use it directly in client-side code unless you fully understand the security risks.

:::
```

**A scoped token can access only those resources that you'll explicitly allow it to.**

### How to create a scoped token

Scoped tokens behave like standard API tokens and are managed through the [Integrations](https://console.apify.com/settings/integrations) page in Apify Console. When creating a token (or updating an existing one), simply toggle "Limit token permissions" to make the token scoped.

Once the token is scoped, you can specify the token's permissions.

![Toggle "Limit token permissions" to make a token scoped](../images/tmp__create-scoped-api-token.png)

### Account level vs resource-specific permissions

We support two different types of permissions for tokens:

- **Account level permissions**: These will apply to all resources in the entire account. For example, you can use these to allow the token to run _all_ your Actors.

- **Resource-specific permissions**: These will apply only to specific, existing resources. For example, you can use these to allow the token to read from a particular dataset.

> :::tip
> A single token can combine both types. You can create a token that can _read_ any data storage, but _write_ only to one specific dataset.
> :::

Reviewer: ```suggestion
:::tip Combined token permissions

A single token can combine both types. You can create a token that can _read_ any data storage, but _write_ only to one specific dataset.

:::
```

Author: Are these callout titles a strong requirement?

TBH, I don't see that much value in them... ideally there'd be no title, just the text plus the icon, but since the title has to be there, for me "Tip" / "Note" / "Important" work better. The proposed titles seem very vague and non-informative to me.

Reviewer: In my experience documentation is more scanned than read. These admonitions and titles within them, can help grab attention as well as at a glance help user decide whether he found what was looking for or not. Moreover that is also how we are doing them anywhere else in docs. I would be in favour of keeping the title, but can definitely come up with few alternatives if these seem to vague

Author: > In my experience documentation is more scanned than read.

Totally on board with this :+1: 

But IMHO the generic titles such as "important" or "tip" can actually greatly improve scannability: If something is a tip, I know I can skip it, if it's important, I should probably read it etc.

My argument is basically that these generic titles might bring more value than callout-specific, yet non-informative titles.

Btw Docusaurus itself [combines both approaches](https://docusaurus.io/docs/next/markdown-features/toc). Specific titles where it makes sense, generic where it doesn't.

I'll also try to think about alternatives for specific titles, but I'd actually prefer to keep the generic ones as fallback.

![An example scoped token that combines account level permissions and resource-specific permissions](../images/tmp__scoped-token-combining-permissions.png)

> :::note Note
> Keep in mind that the scoped tokens still operate within the context of your account. A scoped token will never be able to access resources that your account doesn't have access to.
> :::

Reviewer: ```suggestion
:::note Token scope limits

Scoped tokens operate within the context of your account and cannot access resources beyond your account's permissions.

:::
```

Author: What's the reasoning here? "Keep in mind" was meant to acknowledge the fact that this is actually to some extent obvious ("goes without saying"), so it is basically just a reminder to the reader how this works.

Reviewer: Keep in mind from my POV seems like unnecessary fluff. In admonitions I try to go straight to the point and not make any assumptions about knowledge of reader. To some it might be obvious but not all.

Author: Okay, in that case I'd probably remove this callout completely. It doesn't bring much value anyway.

Author: Removed.

> ### Allowing tokens to create resources

Reviewer: ```suggestion
### Allow tokens to create resources
```

Author: Hmmm here the idea is not to _allow the token to create resource_, but rather _how it works when you allow tokens to create resources_.

If the gerund is a problem, I'd go for something like "How creating resources with scoped tokens work"... but that's a tad long and awkward.

If you need to create new resources with the token (for example, create a new Task, or storage), you need to explicitly allow that as well.

> Once you create a new resource with the token, **the token will gain full access to that resource**, regardless of other permissions. It is not possible to create a token that can create a dataset, but not write to it.

Reviewer: ```suggestion
Once you create a new resource with the token, _the token will gain full access to that resource_, regardless of other permissions. It is not possible to create a token that can create a dataset, but not write to it.
```

> :::tip
> This is useful if you want to for example create a token that can dynamically create & populate datasets, but without the need to access other datasets in your account.
> :::

Reviewer: ```suggestion
:::tip Selective token permissions

This is useful if you want to for example create a token that can dynamically create & populate datasets, but without the need to access other datasets in your account.

:::
```

> ### Running and scheduling Actors

Reviewer: ```suggestion
### Run and schedule Actors
```

Author: Same as above, this is more _how running and scheduling Actors work_ than _how to run and schedule Actors_.

When you run an Actor with a scoped token (or schedule one), Apify will inject a new, _unscoped_ token to the Actor. This means that **the Actor will have full access to all resources in your account**.

This way you can be sure that once you give a token the permission to run an Actor, it will just work, and you don't have to worry
about the exact permissions the Actor might need. However, this also means that you need to trust the Actor.

> :::note Note
> This is why **we currently do not allow scoped tokens to create or modify Actors**. With those permissions it would be easy for the token to upload malicious code and gain access to your full account. If you do need to create or modify Actors via Apify API, you need to use an unscoped token.
> :::

Reviewer: ```suggestion
:::note Actor modification restriction 

For security reasons, scoped tokens cannot create or modify Actors. Use an unscoped token for these operations.

:::
```

Author: The idea here is to make it less wordy? :thinking: 

I'd still like to link the statement here to the previous paragraph.

> ### Setting up webhooks

Reviewer: ```suggestion
### Set up webhooks
```

Author: Dtto as above.

If you allow a token to run an Actor, it'll also be able to manage the Actor's webhooks (similarly for tasks).

If you set up a webhook pointing to the Apify API, the Apify platform will automatically inject a token when dispatching that webhook. However, if you set up such a webhook with a scoped token, **that webhook will be dispatched with a token with the same limited permissions**.

Therefore, you need to make sure the token has sufficient permissions not only to set up the webhook, but also to perform the actual operation.

> :::tip Example
> Let's say you want to create a webhook that pushes an item to a dataset every time an Actor successfully finishes. Then such a scoped token needs to be allowed to both run the Actor (in order to create the webhook), and write to that dataset.
> :::

Reviewer: ```suggestion
:::note Multi-permission example

Let's say you want to create a webhook that pushes an item to a dataset every time an Actor successfully finishes. Then such a scoped token needs to be allowed to both run the Actor (in order to create the webhook), and write to that dataset.

:::
```

:::

### Allowing tokens to create resources

If you need to create new resources with the token (for example, create a new Task, or storage), you need to explicitly allow that as well.

Once you create a new resource with the token, **the token will gain full access to that resource**, regardless of other permissions. It is not possible to create a token that can create a dataset, but not write to it.

:::tip
This is useful if you want to for example create a token that can dynamically create & populate datasets, but without the need to access other datasets in your account.
:::

### Running and scheduling Actors

When you run an Actor with a scoped token (or schedule one), Apify will inject a new, _unscoped_ token to the Actor. This means that **the Actor will have full access to all resources in your account**.

This way you can be sure that once you give a token the permission to run an Actor, it will just work, and you don't have to worry
about the exact permissions the Actor might need. However, this also means that you need to trust the Actor.

> :::info Note

Reviewer 2: It is so confusing that `:::info` has a !⃝ icon, but `:::note` has a ⓘ icon 🤯

Author: I actually did not notice this :smile: Yeah, it's weird... unified it to use `note` across the board. It's less scary.
